chainer/links/loss/hierarchical_softmax.py

```diff
@@ -349,6 +405,79 @@ def create_huffman_tree(word_counts):
 
         return q.get()[2]
 
+    def sample(self, x):
+        """Sample an example for a given input from the tree.
+
+        Args:
+            x (~chainer.Variable): Input variable for sample word ids.
+
+        Returns:
+            array: Array of word indexes in a binary tree ``self.tree``.
+
+        .. admonition:: Example
+
+            Let an input vector ``x`` be:
+
+            >>> word_cnts = {0: 8, 1: 5, 2: 6, 3: 4, 4: 10, 5: 1, 6: 32}
+            >>> tree = BinaryHierarchicalSoftmax.create_huffman_tree(word_cnts)
+            >>> hsm = BinaryHierarchicalSoftmax(3, tree)
+            >>> x = np.array([[0.2, 0.2, 0.3], [0.1, 0.3, 0.1]], dtype='f')
+            >>> hsm.sample(Variable(x))
+            [0, 3]
+
+
+        """
+        if len(self.tree) == 0:
+            raise ValueError('Empty tree')
+
+        xp = cuda.get_array_module(*x)
+        parent2child = self._func.parent2child
+        node2word = self._func.node2word
+        batchsize = x.shape[0]
+        start_ids = xp.zeros(batchsize, 'i')
+        list_next_ids = []
+        list_sampled_ids = []
+
+        def _sigmoid(x):
+            half = x.dtype.type(0.5)
+            return self.xp.tanh(x * half) * half + half
+
+        rows = xp.arange(batchsize, dtype=xp.int32)
+        while True:
+            w = self.W.data[start_ids]
+            x_t = xp.transpose(x.data)
+            score = xp.sum(xp.dot(w, x_t), axis=1)
```

Comment:
Is it really correct?
I think `w` has `(batch, n_unit)` shape, and `x_t` has `(n_unit, batch)` shape. What you want to do is `numpy.einsum('ij,ij->i', w, x)` or `sum(w * x, axis=1)`, right?